Heartbeat RPC getting too big for DB

Resulting in no container getting allocated. (And much more.)

	2015-06-26 11:59:50,235 WARN org.apache.hadoop.ipc.Server: IPC Server handler 23 on 8031, call org.apache.hadoop.yarn.server.api.ResourceTrackerPB.nodeHeartbeat from 193.10.66.125:49380 Call#68001 Retry#0: error: com.mysql.clusterj.ClusterJDatastoreException: For field rpc column rpc valueDelegate setBytes, error executing objectSetValue. Caused by com.mysql.clusterj.ClusterJUserException:Data length 16,703 too long. Column 'rpc' can only accept data of length 13,000.
	com.mysql.clusterj.ClusterJDatastoreException: For field rpc column rpc valueDelegate setBytes, error executing objectSetValue. Caused by com.mysql.clusterj.ClusterJUserException:Data length 16,703 too long. Column 'rpc' can only accept data of length 13,000.
		at com.mysql.clusterj.core.metadata.AbstractDomainFieldHandlerImpl.objectSetValue(AbstractDomainFieldHandlerImpl.java:289)
		at com.mysql.clusterj.tie.NdbRecordSmartValueHandlerImpl.set(NdbRecordSmartValueHandlerImpl.java:724)
		at com.mysql.clusterj.tie.NdbRecordSmartValueHandlerImpl.invoke(NdbRecordSmartValueHandlerImpl.java:750)
		at com.mysql.clusterj.core.metadata.$Proxy31.setrpc(Unknown Source)
		at io.hops.metadata.ndb.dalimpl.yarn.rmstatestore.RPCClusterJ.createPersistable(RPCClusterJ.java:174)
		at io.hops.metadata.ndb.dalimpl.yarn.rmstatestore.RPCClusterJ.add(RPCClusterJ.java:106)
		at io.hops.metadata.ndb.dalimpl.yarn.rmstatestore.RPCClusterJ.add(RPCClusterJ.java:41)
		at io.hops.metadata.util.RMUtilities$21.performTask(RMUtilities.java:756)
		at io.hops.transaction.handler.LightWeightRequestHandler.execute(LightWeightRequestHandler.java:50)
		at io.hops.transaction.handler.RequestHandler.handle(RequestHandler.java:52)
		at io.hops.transaction.handler.RequestHandler.handle(RequestHandler.java:47)
		at io.hops.metadata.util.RMUtilities.persistAppMasterRPC(RMUtilities.java:764)
		at io.hops.metadata.util.RMUtilities.persistAppMasterRPC(RMUtilities.java:741)
		at org.apache.hadoop.yarn.server.resourcemanager.ResourceTrackerService.nodeHeartbeat(ResourceTrackerService.java:459)
		at org.apache.hadoop.yarn.server.resourcemanager.ResourceTrackerService.nodeHeartbeat(ResourceTrackerService.java:443)
		at org.apache.hadoop.yarn.server.api.impl.pb.service.ResourceTrackerPBServiceImpl.nodeHeartbeat(ResourceTrackerPBServiceImpl.java:66)
		at org.apache.hadoop.yarn.proto.ResourceTracker$ResourceTrackerService$2.callBlockingMethod(ResourceTracker.java:81)
		at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:585)
		at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:928)
		at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2013)
		at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2009)
		at java.security.AccessController.doPrivileged(Native Method)
		at javax.security.auth.Subject.doAs(Subject.java:415)
		at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1548)
		at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2007)
	Caused by: com.mysql.clusterj.ClusterJUserException: Data length 16,703 too long. Column 'rpc' can only accept data of length 13,000.
		at com.mysql.clusterj.tie.NdbRecordImpl.setBytes(NdbRecordImpl.java:300)
		at com.mysql.clusterj.tie.NdbRecordOperationImpl.setBytes(NdbRecordOperationImpl.java:474)
		at com.mysql.clusterj.tie.NdbRecordOperationImpl.setBytes(NdbRecordOperationImpl.java:480)
		at com.mysql.clusterj.tie.NdbRecordSmartValueHandlerImpl.setBytes(NdbRecordSmartValueHandlerImpl.java:496)
		at com.mysql.clusterj.core.metadata.AbstractDomainFieldHandlerImpl$6.objectSetValue(AbstractDomainFieldHandlerImpl.java:858)
		at com.mysql.clusterj.core.metadata.AbstractDomainFieldHandlerImpl.objectSetValue(AbstractDomainFieldHandlerImpl.java:287)
		... 24 more

--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/io/hops/metadata/util/RMUtilities.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/io/hops/metadata/util/RMUtilities.java
@@ -15,7 +15,6 @@
  */
 package io.hops.metadata.util;
 
-import com.google.common.util.concurrent.AtomicDouble;
 import com.google.protobuf.InvalidProtocolBufferException;
 import io.hops.exception.StorageException;
 import io.hops.ha.common.TransactionState;
@@ -50,18 +49,19 @@ import io.hops.metadata.yarn.dal.ResourceDataAccess;
 import io.hops.metadata.yarn.dal.ResourceRequestDataAccess;
 import io.hops.metadata.yarn.dal.SchedulerApplicationDataAccess;
 import io.hops.metadata.yarn.dal.UpdatedContainerInfoDataAccess;
-import io.hops.metadata.yarn.dal.YarnVariablesDataAccess;
 import io.hops.metadata.yarn.dal.capacity.CSLeafQueueUserInfoDataAccess;
 import io.hops.metadata.yarn.dal.capacity.CSQueueDataAccess;
 import io.hops.metadata.yarn.dal.capacity.FiCaSchedulerAppReservedContainersDataAccess;
 import io.hops.metadata.yarn.dal.fair.AppSchedulableDataAccess;
 import io.hops.metadata.yarn.dal.fair.FSSchedulerNodeDataAccess;
+import io.hops.metadata.yarn.dal.rmstatestore.AllocateRPCDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.AllocateResponseDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.AllocatedContainersDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.ApplicationAttemptStateDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.ApplicationStateDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.DelegationKeyDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.DelegationTokenDataAccess;
+import io.hops.metadata.yarn.dal.rmstatestore.HeartBeatRPCDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.RMStateVersionDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.RPCDataAccess;
 import io.hops.metadata.yarn.dal.rmstatestore.RanNodeDataAccess;
@@ -96,7 +96,8 @@ import io.hops.metadata.yarn.entity.ResourceRequest;
 import io.hops.metadata.yarn.entity.SchedulerAppReservations;
 import io.hops.metadata.yarn.entity.SchedulerApplication;
 import io.hops.metadata.yarn.entity.UpdatedContainerInfo;
-import io.hops.metadata.yarn.entity.YarnVariables;
+import io.hops.metadata.yarn.entity.appmasterrpc.AllocateRPC;
+import io.hops.metadata.yarn.entity.appmasterrpc.HeartBeatRPC;
 import io.hops.metadata.yarn.entity.appmasterrpc.RPC;
 import io.hops.metadata.yarn.entity.capacity.CSLeafQueueUserInfo;
 import io.hops.metadata.yarn.entity.capacity.CSQueue;
@@ -444,6 +445,20 @@ public class RMUtilities {
     return DA.getAll();
   }
 
+  public static Map<Integer, HeartBeatRPC> getHeartBeatRPCs() throws
+          StorageException {
+    HeartBeatRPCDataAccess hbDA = (HeartBeatRPCDataAccess) RMStorageFactory.
+            getDataAccess(HeartBeatRPCDataAccess.class);
+    return hbDA.getAll();
+  }
+  
+  public static Map<Integer, AllocateRPC> getAllocateRPCs() throws
+          StorageException {
+    AllocateRPCDataAccess alDA = (AllocateRPCDataAccess) RMStorageFactory.
+            getDataAccess(AllocateRPCDataAccess.class);
+    return alDA.getAll();
+  }
+  
   public static List<PendingEvent> getAllPendingEvents() throws IOException {
 
     PendingEventDataAccess DA = (PendingEventDataAccess) RMStorageFactory.
@@ -750,131 +765,62 @@ public static Map<String, List<ResourceRequest>> getAllResourceRequestsFullTrans
     setAppMasterRPCHandler.handle();
   }
 
-
-  /**
-   * Checks for pending RMNode RPCs and if the RMNode is active.
-   * If an rpc for a particular RMNode and with the same type already exists
-   * the method returns false. If the RPC does not exist it checks if the
-   * RMNode is in the ActiveNodes table. If so, it returns the retrieved
-   * RMNode.
-   * <p/>
-   *
-   * @param type
-   * @param rpc
-   * @param userId
-   * @return
-   * @throws IOException
-   */
-  public static Map<Integer, Object> registerNMRPCValidation(
-      final RPC.Type type, final byte[] rpc, final String userId)
-      throws IOException {
-    LightWeightRequestHandler registerNMRPCValidationHandler =
-        new LightWeightRequestHandler(YARNOperationType.TEST) {
-          @Override
-          public Object performTask() throws StorageException {
-            connector.beginTransaction();
-            connector.writeLock();
-            Map<Integer, Object> toReturn;
-            //Check if an rpc of the same type for this RMNode already exists
-            RPCDataAccess rpcDA = (RPCDataAccess) RMStorageFactory
+  public static void persistHeartBeatRPC(final HeartBeatRPC rpc) throws
+          IOException {
+    LightWeightRequestHandler heartBeatRPCHandler
+            = new LightWeightRequestHandler(YARNOperationType.TEST) {
+              @Override
+              public Object performTask() throws StorageException {
+                connector.beginTransaction();
+                connector.writeLock();
+                RPCDataAccess DA = (RPCDataAccess) RMStorageFactory
                 .getDataAccess(RPCDataAccess.class);
-            //If it exists, return false
-            if (rpcDA.findByTypeAndUserId(type.toString(), userId) /*||
-                        rpcDA.findByTypeAndUserId(HopRPC.Type.NodeHeartbeat.toString(), userId)*/) {
-              LOG.debug("HOP :: rmnodeRPCValidation() - RPC already exists");
-              connector.commit();
-              return null;
-            } else {
-              LOG.debug("HOP :: rmnodeRPCValidation() rpcIdFound was null");
-              //Get new rpcId and persist it
-              YarnVariablesDataAccess yDA =
-                  (YarnVariablesDataAccess) RMStorageFactory
-                      .getDataAccess(YarnVariablesDataAccess.class);
-              YarnVariables found =
-                  (YarnVariables) yDA.findById(HopYarnAPIUtilities.RPC);
-              int rpcId = Integer.MIN_VALUE;
-              if (found != null) {
-                rpcId = found.getValue();
-                found.setValue(found.getValue() + 1);
-                yDA.add(found);
+                RPC hop = new RPC(rpc.getRpcId(), RPC.Type.NodeHeartbeat, null,
+                        null);
+                DA.add(hop);
+                connector.flush();
+                HeartBeatRPCDataAccess hbDA
+                = (HeartBeatRPCDataAccess) RMStorageFactory.getDataAccess(
+                        HeartBeatRPCDataAccess.class);
+
+                hbDA.add(rpc);
+
+
+                connector.commit();
+                return null;
               }
-              RPC rpcToPersist = new RPC(rpcId, type, rpc, userId);
-              rpcDA.add(rpcToPersist);
-              toReturn = new HashMap<Integer, Object>();
-              toReturn.put(0, rpcId);
-            }
-            //Check if RMNode(userId) is in ActiveRMNodesMap
-            RMContextActiveNodesDataAccess rmDA =
-                (RMContextActiveNodesDataAccess) RMStorageFactory.
-                    getDataAccess(RMContextActiveNodesDataAccess.class);
-
-            RMContextActiveNodes hopActiveRMNode =
-                (RMContextActiveNodes) rmDA.findEntry(userId);
-            if (hopActiveRMNode != null) {
-              toReturn.put(1, true);
-            } else {
-              toReturn.put(1, false);
-            }
-            connector.commit();
-            return toReturn;
-          }
-        };
-    return (Map<Integer, Object>) registerNMRPCValidationHandler.handle();
+            };
+    heartBeatRPCHandler.handle();
   }
+  
+  public static void persistAllocateRPC(final AllocateRPC rpc,
+          final String userId) throws IOException {
+    LightWeightRequestHandler allocateRPCHandler
+            = new LightWeightRequestHandler(YARNOperationType.TEST) {
 
-  /**
-   * Checks for pending heartbeats.
-   *
-   * @param type
-   * @param rpc
-   * @param id
-   * @param rmContext
-   * @return
-   * @throws IOException
-   */
-  public static int heartbeatNMRPCValidation(final RPC.Type type,
-      final byte[] rpc, final String id, final RMContext rmContext)
-      throws IOException {
-    LightWeightRequestHandler heartbeatNMRPCValidationHandler =
-        new LightWeightRequestHandler(YARNOperationType.TEST) {
-          @Override
-          public Object performTask() throws StorageException {
-            LOG.debug("HOP :: heartbeatNMRPCValidation - START");
-            connector.beginTransaction();
-            connector.writeLock();
-            //Check if an rpc of the same type for this RMNode already exists
-            RPCDataAccess rpcDA = (RPCDataAccess) RMStorageFactory
+              @Override
+              public Object performTask() throws IOException {
+                connector.beginTransaction();
+                connector.writeLock();
+                RPCDataAccess DA = (RPCDataAccess) RMStorageFactory
                 .getDataAccess(RPCDataAccess.class);
-            int rpcId = Integer.MIN_VALUE;
-            if (rpcDA.findByTypeAndUserId(type.toString(), id)) {
-              LOG.debug(
-                  "HOP :: heartbeatNMRPCValidation() - RPC already exists");
-              connector.commit();
-              return Integer.MIN_VALUE;
-            } else {
-              //Get new rpcId and persist it
-              YarnVariablesDataAccess yDA =
-                  (YarnVariablesDataAccess) RMStorageFactory
-                      .getDataAccess(YarnVariablesDataAccess.class);
-              YarnVariables found =
-                  (YarnVariables) yDA.findById(HopYarnAPIUtilities.RPC);
-              if (found != null) {
-                rpcId = found.getValue();
-                found.setValue(found.getValue() + 1);
+                RPC hop = new RPC(rpc.getRpcID(), RPC.Type.Allocate, null,
+                        userId);
+                DA.add(hop);
+                connector.flush();
+                AllocateRPCDataAccess aDA
+                = (AllocateRPCDataAccess) RMStorageFactory.getDataAccess(
+                        AllocateRPCDataAccess.class);
 
-                yDA.add(found);
+                aDA.add(rpc);
+
+                connector.commit();
+                return null;
               }
-              RPC rpcToPersist = new RPC(rpcId, type, rpc, id);
-              rpcDA.add(rpcToPersist);
-            }
-            connector.commit();
-            LOG.debug("HOP :: heartbeatNMRPCValidation - FINISH");
-            return rpcId;
-          }
-        };
-    return (Integer) heartbeatNMRPCValidationHandler.handle();
+            };
+    allocateRPCHandler.handle();
   }
-
+  
   public static void updatePendingEvents(final PendingEvent persistedEvent,
       final int action) throws IOException {
     LightWeightRequestHandler updatePendingEventsHandler =

--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ApplicationMasterService.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ApplicationMasterService.java
@@ -18,10 +18,10 @@ package org.apache.hadoop.yarn.server.resourcemanager;
 
 import com.google.common.annotations.VisibleForTesting;
 import io.hops.ha.common.TransactionState;
-import io.hops.ha.common.TransactionState.TransactionType;
 import io.hops.ha.common.TransactionStateImpl;
 import io.hops.metadata.util.HopYarnAPIUtilities;
 import io.hops.metadata.util.RMUtilities;
+import io.hops.metadata.yarn.entity.appmasterrpc.AllocateRPC;
 import io.hops.metadata.yarn.entity.appmasterrpc.RPC;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
@@ -92,12 +92,16 @@ import java.net.InetSocketAddress;
 import java.net.UnknownHostException;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import org.apache.hadoop.yarn.api.records.ContainerResourceIncreaseRequest;
+import org.apache.hadoop.yarn.api.records.impl.pb.ContainerResourceIncreaseRequestPBImpl;
+import org.apache.hadoop.yarn.api.records.impl.pb.ResourceRequestPBImpl;
 import org.apache.hadoop.yarn.server.resourcemanager.recovery.RMStateStore;
 
 @SuppressWarnings("unchecked")
@@ -461,18 +465,47 @@ public class ApplicationMasterService extends AbstractService
     if (rpcID == null) {
       rpcID = HopYarnAPIUtilities.getRPCID();
       byte[] allAMRequestData = ((AllocateRequestPBImpl) request).getProto().
-          toByteArray();
-      if(allAMRequestData.length>13000){
-        //TORECOVER persist in a prorper way
-        LOG.error("RPC Too big: " + request.getReleaseList().size() + ", " + request.getAskList().size() + ", " + request.getIncreaseRequests().size());
-        AllocateRequest substitut = AllocateRequest.newInstance(request.getResponseId(),
-                request.getProgress(), new ArrayList<ResourceRequest>(), new ArrayList<ContainerId>(), request.getResourceBlacklistRequest(), new ArrayList<ContainerResourceIncreaseRequest>());
-        allAMRequestData = ((AllocateRequestPBImpl)substitut).getProto().toByteArray();
+              toByteArray();
+
+      List<String> releaseList = new ArrayList<String>();
+      for (ContainerId containerId : request.getReleaseList()) {
+        releaseList.add(containerId.toString());
       }
+
+      Map<String, byte[]> ask = new HashMap<String, byte[]>();
+      for (ResourceRequest resourceRequest : request.getAskList()) {
+        resourceRequest.getCapability();
+
+        ask.put(resourceRequest.getResourceName(),
+                ((ResourceRequestPBImpl) resourceRequest).getProto().
+                toByteArray());
+      }
+
+      Map<String, byte[]> resourceIncreaseRequest
+              = new HashMap<String, byte[]>();
+      for (ContainerResourceIncreaseRequest incRequest : request.
+              getIncreaseRequests()) {
+        resourceIncreaseRequest.put(incRequest.getContainerId().toString(),
+                ((ContainerResourceIncreaseRequestPBImpl) incRequest).getProto().
+                toByteArray());
+      }
+
+      List<String> blackListAddition = new ArrayList<String>();
+      List<String> blackListRemovals = new ArrayList<String>();
+      if (request.getResourceBlacklistRequest() != null) {
+        blackListAddition.addAll(request.getResourceBlacklistRequest().
+                getBlacklistAdditions());
+
+        blackListRemovals.addAll(request.getResourceBlacklistRequest().
+                getBlacklistRemovals());
+      }
+      AllocateRPC rpc = new AllocateRPC(rpcID, request.getResponseId(), request.
+              getProgress(), releaseList, ask, resourceIncreaseRequest,
+              blackListAddition, blackListRemovals);
+
       RMUtilities
-          .persistAppMasterRPC(rpcID, RPC.Type.Allocate, allAMRequestData,
-              appAttemptId.toString());
-      
+              .persistAllocateRPC(rpc, appAttemptId.toString());
+
     }
     TransactionState transactionState = 
            rmContext.getTransactionStateManager().getCurrentTransactionStateNonPriority(rpcID,

--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ResourceManager.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ResourceManager.java
@@ -22,6 +22,8 @@ import io.hops.ha.common.TransactionStateManager;
 import io.hops.metadata.util.RMStorageFactory;
 import io.hops.metadata.util.YarnAPIStorageFactory;
 import io.hops.metadata.yarn.entity.PendingEvent;
+import io.hops.metadata.yarn.entity.appmasterrpc.AllocateRPC;
+import io.hops.metadata.yarn.entity.appmasterrpc.HeartBeatRPC;
 import io.hops.metadata.yarn.entity.appmasterrpc.RPC;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
@@ -118,10 +120,25 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.net.InetSocketAddress;
 import java.security.PrivilegedExceptionAction;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.LinkedBlockingQueue;
+import org.apache.hadoop.yarn.api.records.ContainerId;
+import org.apache.hadoop.yarn.api.records.ContainerResourceIncreaseRequest;
+import org.apache.hadoop.yarn.api.records.ContainerStatus;
+import org.apache.hadoop.yarn.api.records.ResourceRequest;
+import org.apache.hadoop.yarn.api.records.impl.pb.ContainerResourceIncreaseRequestPBImpl;
+import org.apache.hadoop.yarn.api.records.impl.pb.ContainerStatusPBImpl;
+import org.apache.hadoop.yarn.api.records.impl.pb.NodeIdPBImpl;
+import org.apache.hadoop.yarn.api.records.impl.pb.ResourceBlacklistRequestPBImpl;
+import org.apache.hadoop.yarn.api.records.impl.pb.ResourceRequestPBImpl;
+import org.apache.hadoop.yarn.proto.YarnProtos;
+import org.apache.hadoop.yarn.proto.YarnServerCommonProtos;
+import org.apache.hadoop.yarn.server.api.records.impl.pb.MasterKeyPBImpl;
+import org.apache.hadoop.yarn.server.api.records.impl.pb.NodeHealthStatusPBImpl;
+import org.apache.hadoop.yarn.server.api.records.impl.pb.NodeStatusPBImpl;
 import org.apache.hadoop.yarn.server.resourcemanager.scheduler.quota.QuotaSchedulerService;
 
 /**
@@ -1275,7 +1292,8 @@ public class ResourceManager extends CompositeService implements Recoverable {
     }
   }
   
-  protected void recoverRpc(RMState rmState) throws IOException, YarnException {
+  protected void recoverRpc(final RMState rmState) throws IOException,
+          YarnException {
     List<RPC> rpcList = rmState.getAppMasterRPCs();
     if(!rpcList.isEmpty()){
       Collections.sort(rpcList);
@@ -1329,16 +1347,55 @@ public class ResourceManager extends CompositeService implements Recoverable {
             ugi = creatAMRMTokenUGI(rpc);
             ugi.doAs(new PrivilegedExceptionAction<AllocateResponse>() {
 
-                  @Override
-                  public AllocateResponse run() throws Exception {
-                    com.google.protobuf.GeneratedMessage proto =
-                        YarnServiceProtos.AllocateRequestProto.parseFrom(rpc.
-                            getRpc());
-                    return masterService.allocate(new AllocateRequestPBImpl(
-                            (YarnServiceProtos.AllocateRequestProto) proto),
-                        rpc.getRPCId());
-                  }
-                });
+              @Override
+              public AllocateResponse run() throws Exception {
+
+                AllocateRequestPBImpl request = new AllocateRequestPBImpl();
+
+                AllocateRPC allocateRPC = rmState.getAllocateRPCs().get(rpc.
+                        getRPCId());
+
+                List<ResourceRequest> askList
+                        = new ArrayList<ResourceRequest>();
+                for (byte[] ask : allocateRPC.getAsk().values()) {
+                  askList.add(new ResourceRequestPBImpl(
+                          YarnProtos.ResourceRequestProto.parseFrom(
+                                  ask)));
+                }
+                request.setAskList(askList);
+
+                List<ContainerResourceIncreaseRequest> incRequestList
+                        = new ArrayList<ContainerResourceIncreaseRequest>();
+                for (byte[] incRequest : allocateRPC.
+                        getResourceIncreaseRequest().values()) {
+                  incRequestList.add(new ContainerResourceIncreaseRequestPBImpl(
+                          YarnProtos.ContainerResourceIncreaseRequestProto.
+                          parseFrom(incRequest)));
+                }
+                request.setIncreaseRequests(incRequestList);
+
+                request.setProgress(allocateRPC.getProgress());
+
+                List<ContainerId> releaseList = new ArrayList<ContainerId>();
+                for (String containerId : allocateRPC.getReleaseList()) {
+                  releaseList.add(ConverterUtils.toContainerId(containerId));
+                }
+                request.setReleaseList(releaseList);
+
+                ResourceBlacklistRequestPBImpl blackListRequest
+                        = new ResourceBlacklistRequestPBImpl();
+
+                blackListRequest.setBlacklistAdditions(allocateRPC.
+                        getBlackListAddition());
+                blackListRequest.setBlacklistRemovals(allocateRPC.
+                        getBlackListRemovals());
+                request.setResourceBlacklistRequest(blackListRequest);
+
+                request.setResponseId(allocateRPC.getResponseId());
+
+                return masterService.allocate(request, rpc.getRPCId());
+              }
+            });
             break;
           case SubmitApplication:
             ugi = UserGroupInformation.createRemoteUser(rpc.getUserId());
@@ -1387,12 +1444,54 @@ public class ResourceManager extends CompositeService implements Recoverable {
             }
             break;
           case NodeHeartbeat:
-            if (!rmContext.isDistributedEnabled() || rmContext.getGroupMembershipService().isAlone()) {
-              proto = YarnServerCommonServiceProtos.NodeHeartbeatRequestProto.
-                      parseFrom(rpc.getRpc());
-              resourceTracker.nodeHeartbeat(new NodeHeartbeatRequestPBImpl(
-                      (YarnServerCommonServiceProtos.NodeHeartbeatRequestProto) proto),
-                      rpc.getRPCId());
+            if (!rmContext.isDistributedEnabled() || rmContext.
+                    getGroupMembershipService().isAlone()) {
+              NodeHeartbeatRequestPBImpl heartBeatRPC
+                      = new NodeHeartbeatRequestPBImpl();
+              HeartBeatRPC hbRPC = rmState.getHeartBeatRPCs().
+                      get(rpc.getRPCId());
+              YarnServerCommonProtos.MasterKeyProto mkProto
+                      = YarnServerCommonProtos.MasterKeyProto.parseFrom(hbRPC.
+                              getLastKnownContainerTokenMasterKey());
+              heartBeatRPC.setLastKnownContainerTokenMasterKey(
+                      new MasterKeyPBImpl(mkProto));
+
+              mkProto = YarnServerCommonProtos.MasterKeyProto.parseFrom(hbRPC.
+                      getLastKnownNMTokenMasterKey());
+              heartBeatRPC.setLastKnownNMTokenMasterKey(new MasterKeyPBImpl(
+                      mkProto));
+
+              NodeStatusPBImpl nodeStatus = new NodeStatusPBImpl();
+
+              List<ContainerStatus> containersStatuses
+                      = new ArrayList<ContainerStatus>();
+              for (byte[] statusBytes
+                      : hbRPC.getContainersStatuses().values()) {
+                YarnProtos.ContainerStatusProto csProto
+                        = YarnProtos.ContainerStatusProto.parseFrom(
+                                statusBytes);
+                containersStatuses.add(new ContainerStatusPBImpl(csProto));
+              }
+              nodeStatus.setContainersStatuses(containersStatuses);
+
+              List<ApplicationId> keepAliveApps = new ArrayList<ApplicationId>();
+              for (String appId : hbRPC.getKeepAliveApplications()) {
+                keepAliveApps.add(ConverterUtils.toApplicationId(appId));
+              }
+              nodeStatus.setKeepAliveApplications(keepAliveApps);
+
+              YarnServerCommonProtos.NodeHealthStatusProto nhProto
+                      = YarnServerCommonProtos.NodeHealthStatusProto.parseFrom(
+                              hbRPC.getNodeHealthStatus());
+              nodeStatus.
+                      setNodeHealthStatus(new NodeHealthStatusPBImpl(nhProto));
+
+              nodeStatus.setNodeId(ConverterUtils.toNodeId(hbRPC.getNodeId()));
+              nodeStatus.setResponseId(hbRPC.getResponseId());
+
+              heartBeatRPC.setNodeStatus(nodeStatus);
+
+              resourceTracker.nodeHeartbeat(heartBeatRPC, rpc.getRPCId());
             }
             break;
           default:

--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ResourceTrackerService.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ResourceTrackerService.java
@@ -23,6 +23,7 @@ import io.hops.ha.common.transactionStateWrapper;
 import io.hops.metadata.util.HopYarnAPIUtilities;
 import io.hops.metadata.util.RMStorageFactory;
 import io.hops.metadata.util.RMUtilities;
+import io.hops.metadata.yarn.entity.appmasterrpc.HeartBeatRPC;
 import io.hops.metadata.yarn.entity.appmasterrpc.RPC;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
@@ -75,6 +76,15 @@ import org.apache.hadoop.yarn.util.YarnVersionInfo;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.yarn.api.records.ApplicationId;
+import org.apache.hadoop.yarn.api.records.impl.pb.ContainerStatusPBImpl;
+import org.apache.hadoop.yarn.server.api.records.impl.pb.MasterKeyPBImpl;
+import org.apache.hadoop.yarn.server.api.records.impl.pb.NodeHealthStatusPBImpl;
 
 public class ResourceTrackerService extends AbstractService
         implements ResourceTracker {
@@ -494,9 +504,55 @@ public class ResourceTrackerService extends AbstractService
       rpcID = HopYarnAPIUtilities.getRPCID();
       byte[] allHBRequestData = ((NodeHeartbeatRequestPBImpl) request).
               getProto().toByteArray();
+      
+      Map<String, byte[]> containersStatuses = new HashMap<String, byte[]>();
+      for (ContainerStatus status : request.getNodeStatus().
+              getContainersStatuses()) {
+        containersStatuses.put(status.getContainerId().toString(),
+                ((ContainerStatusPBImpl) status).getProto().toByteArray());
+      }
+
+      List<String> keepAliveApplications = new ArrayList<String>();
+      for (ApplicationId appId : request.getNodeStatus().
+              getKeepAliveApplications()) {
+        keepAliveApplications.add(appId.toString());
+      }
+
+      byte[] nodeHealthStatus = ((NodeHealthStatusPBImpl) request.
+              getNodeStatus().getNodeHealthStatus()).getProto().toByteArray();
+      if (nodeHealthStatus.length > 1000) {
+        LOG.warn("Node Health Status too big for node " + request.
+                getNodeStatus().getNodeId() + " truncating it");
+        NodeHealthStatusPBImpl healthStatus = new NodeHealthStatusPBImpl();
+        healthStatus.setIsNodeHealthy(request.getNodeStatus().
+                getNodeHealthStatus().getIsNodeHealthy());
+        healthStatus.setLastHealthReportTime(request.getNodeStatus().
+                getNodeHealthStatus().getLastHealthReportTime());
+        String healthReport = StringUtils.abbreviate(request.getNodeStatus().
+                getNodeHealthStatus().getHealthReport(), 200);
+        healthStatus.setHealthReport(healthReport);
+        nodeHealthStatus = healthStatus.getProto().toByteArray();
+      }
+
+      byte[] lastKnownContainerTokenMasterKey = null;
+      byte[] lastKnownNMTokenMasterKey = null;
+      if (request.getLastKnownContainerTokenMasterKey() != null) {
+        lastKnownContainerTokenMasterKey = ((MasterKeyPBImpl) request.
+                getLastKnownContainerTokenMasterKey()).getProto().toByteArray();
+      }
+      if (request.getLastKnownNMTokenMasterKey() != null) {
+        lastKnownNMTokenMasterKey = ((MasterKeyPBImpl) request.
+                getLastKnownNMTokenMasterKey()).getProto().toByteArray();
+      }
+      HeartBeatRPC rpc = new HeartBeatRPC(request.getNodeStatus().getNodeId().
+              toString(),
+              request.getNodeStatus().getResponseId(), containersStatuses,
+              keepAliveApplications,
+              nodeHealthStatus, lastKnownContainerTokenMasterKey,
+              lastKnownNMTokenMasterKey, rpcID);
+
       RMUtilities
-              .persistAppMasterRPC(rpcID, RPC.Type.NodeHeartbeat,
-                      allHBRequestData);
+              .persistHeartBeatRPC(rpc);
     }
     TransactionState transactionState = rmContext.getTransactionStateManager().
             getCurrentTransactionStatePriority(rpcID, "nodeHeartbeat");

--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/recovery/NDBRMStateStore.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/recovery/NDBRMStateStore.java
@@ -22,6 +22,7 @@ import io.hops.metadata.yarn.entity.AppSchedulingInfo;
 import io.hops.metadata.yarn.entity.ContainerId;
 import io.hops.metadata.yarn.entity.FinishedApplications;
 import io.hops.metadata.yarn.entity.NodeHBResponse;
+import io.hops.metadata.yarn.entity.appmasterrpc.HeartBeatRPC;
 import io.hops.metadata.yarn.entity.rmstatestore.UpdatedNode;
 import io.hops.metadata.yarn.entity.rmstatestore.DelegationKey;
 import io.hops.metadata.yarn.entity.rmstatestore.DelegationToken;
@@ -336,6 +337,8 @@ public class NDBRMStateStore extends RMStateStore {
    
   private void loadRPCs(RMState rmState) throws IOException {
     rmState.appMasterRPCs = RMUtilities.getAppMasterRPCs();
+    rmState.heartBeatRPCs = RMUtilities.getHeartBeatRPCs();
+    rmState.allocateRPCs = RMUtilities.getAllocateRPCs();
   }
   
   private void loadPendingEvents(RMState rmState) throws IOException{

--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/recovery/RMStateStore.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/recovery/RMStateStore.java
@@ -41,6 +41,8 @@ import io.hops.metadata.yarn.entity.ResourceRequest;
 import io.hops.metadata.yarn.entity.SchedulerAppReservations;
 import io.hops.metadata.yarn.entity.SchedulerApplication;
 import io.hops.metadata.yarn.entity.UpdatedContainerInfo;
+import io.hops.metadata.yarn.entity.appmasterrpc.AllocateRPC;
+import io.hops.metadata.yarn.entity.appmasterrpc.HeartBeatRPC;
 import io.hops.metadata.yarn.entity.appmasterrpc.RPC;
 import io.hops.metadata.yarn.entity.capacity.CSLeafQueueUserInfo;
 import io.hops.metadata.yarn.entity.capacity.CSQueue;
@@ -381,6 +383,8 @@ import org.apache.hadoop.yarn.server.resourcemanager.rmcontainer.RMContainerStat
 
     Map<KeyType, MasterKey> secretMamagerKeys;
     List<RPC> appMasterRPCs;
+    Map<Integer, HeartBeatRPC> heartBeatRPCs;
+    Map<Integer, AllocateRPC> allocateRPCs;
     List<PendingEvent> pendingEvents;
     Map<String, AppSchedulingInfo> appSchedulingInfos;
     Map<String, SchedulerApplication> schedulerApplications;
@@ -447,6 +451,14 @@ import org.apache.hadoop.yarn.server.resourcemanager.rmcontainer.RMContainerStat
       }
     }
 
+    public Map<Integer, HeartBeatRPC> getHeartBeatRPCs(){
+      return heartBeatRPCs;
+    }
+    
+    public Map<Integer, AllocateRPC> getAllocateRPCs(){
+      return allocateRPCs;
+    }
+    
     public List<PendingEvent> getPendingEvents() throws IOException {
       if (pendingEvents != null) {
         return pendingEvents;

--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/TestDistributedRTClient.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/TestDistributedRTClient.java
@@ -387,45 +387,6 @@ public class TestDistributedRTClient {
     LOG.info("HOP :: rmNode.uci-" + ((RMNodeImpl) rmNode).getQueue());
   }
 
-  @Ignore
-  @Test
-  public void testHeartBeatValidation()
-      throws IOException, InterruptedException {
-    RMUtilities.InitializeDB();
-    new Thread() {
-      @Override
-      public void run() {
-        for (int i = 0; i < 10000; i++) {
-          try {
-            LOG.info("HOP :: testHeartBeatValidation A - START");
-            RMUtilities.heartbeatNMRPCValidation(RPC.Type.NodeHeartbeat,
-                new byte[]{0x0}, "host:1234", null);
-            LOG.info("HOP :: testHeartBeatValidation A - FINISH");
-          } catch (IOException ex) {
-            Logger.getLogger(TestDistributedRT.class.getName()).
-                log(Level.SEVERE, null, ex);
-          }
-        }
-      }
-    }.start();
-    new Thread() {
-      @Override
-      public void run() {
-        for (int i = 0; i < 10000; i++) {
-          try {
-            LOG.info("HOP :: testHeartBeatValidation B - START");
-            RMUtilities.getRMNode("h1:1234", null, conf);
-            LOG.info("HOP :: testHeartBeatValidation B - FINISH");
-          } catch (IOException ex) {
-            Logger.getLogger(TestDistributedRT.class.getName()).
-                log(Level.SEVERE, null, ex);
-          }
-        }
-      }
-    }.start();
-
-    Thread.sleep(50000);
-  }
 
   private class RTClientWorker implements Runnable {
 

